{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5565d554-c639-4de3-9e71-de2ddd9da533",
   "metadata": {},
   "source": [
    "# Introduction to TRGTdb API\n",
    "\n",
    "This tutorial will show you how to programmatically access TRGT results via the TRGT API.\n",
    "\n",
    "By the end of this introduction, you will know:\n",
    "- The TRGTdb tables and their structure\n",
    "- How to access the in-memory TRGTdb pandas DataFrames\n",
    "- How to intersect/join across tables\n",
    "- How to use some of the TRGTdb standard queries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4242ed6b-5265-42b4-a34f-20ef7cea9bc5",
   "metadata": {},
   "source": [
    "# Loading a TRGTdb\n",
    "\n",
    "First, we'll want to import trgt and load our database. This assumes that you've already performed a conversion on a TRGT result e.g.\n",
    "```bash\n",
    "trgt db create -o all_vcf.tdb son.vcf.gz mother.vcf.gz father.vcf.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "67595937-3e20-4ca1-a57a-a759957eb840",
   "metadata": {},
   "outputs": [],
   "source": [
    "import trgt\n",
    "\n",
    "data = trgt.load_tdb(\"../test_files/databases/all_vcf.tdb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34b25d10-a956-49a2-9900-5a4ad2ae0d52",
   "metadata": {},
   "source": [
    "This returns a dictionary with structure:\n",
    "\n",
    "| Key | Details |\n",
    "| --- | ------- |\n",
    "| locus | Tandem repeat loci inside the database |\n",
    "| allele | Alleles observed on each locus |\n",
    "| sample | Tables of each sample's observed alleles |\n",
    "\n",
    "## The Locus Table\n",
    "Looking at our `locus` table first, we see that it holds a Pandas DataFrame of the LocusID and genomic position TRGT analyzed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "62e9b39c-1be7-4cbc-9bda-f68cfe8c6e8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16682</td>\n",
       "      <td>16774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19275</td>\n",
       "      <td>19473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20798</td>\n",
       "      <td>20893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29714</td>\n",
       "      <td>29822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   LocusID chrom  start    end\n",
       "0        3  chr1  16682  16774\n",
       "1        9  chr1  19275  19473\n",
       "2        4  chr1  20798  20893\n",
       "3        2  chr1  29714  29822\n",
       "4        0  chr1  30824  30989"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\"locus\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0692ef20-f092-464e-bdb0-a373832ead81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of loci: 77213\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total number of loci: {len(data['locus'])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe3a7b7-3e86-446b-9494-6e1262510376",
   "metadata": {},
   "source": [
    "Because the table in-memory is just a DataFrame, we can of course do analyis on it. For example, if we wanted to know how many base-pairs are spanned by tandem repeat loci:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1e34484-543e-4493-b8bf-f6e8a9304375",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total bases: 9482540\n"
     ]
    }
   ],
   "source": [
    "loci = data[\"locus\"] # Make a direct variable to the table so we don't have to type `data[\"locus\"]` every time\n",
    "locus_span = loci['end'] - loci['start']\n",
    "total_locus_span = locus_span.sum()\n",
    "print(f\"Total bases: {total_locus_span}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d71115f-30f3-424c-93c0-c110eba8bba2",
   "metadata": {},
   "source": [
    "Similarly, we can use the `locus_span` to check the distribution lengths. We'll use seaborn to plot a histogram."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7cee4e00-f400-45aa-976d-bc26ee7cbcc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAloAAAHHCAYAAABnS/bqAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA+eUlEQVR4nO3df3zO9f7H8ee1sR/MNst+WOa3SH6FmpUfyTLaqdROJ3Jq4qhkKiuJ8rMcRY4UcZxOdM5JSkWFZH5NZRQNEcJRiCFsl58z2/v7h7PP12Ubu7SP7ZrH/Xa7brddn/fr+lyv9z7Tnn0+7+szhzHGCAAAACXOq7QbAAAAKK8IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAHAF1K5dW7169bL9fX7++Wc5HA7NnDnT2tarVy8FBATY/t75HA6HRo4cecXeDyjLCFqAB5g5c6YcDofWrl1b2q2UqM8//1wdOnRQWFiYKlWqpLp16+pPf/qTFi1aVNqtXdRtt90mh8Mhh8MhLy8vBQYGqmHDhnrooYeUkpJSYu+zcOHCMhtYynJvQFlSobQbAHB1eu211zRo0CB16NBBQ4YMUaVKlbRjxw4tWbJEs2fPVpcuXUq7xYuqUaOGxo4dK0k6ceKEduzYoU8++UT/+c9/9Kc//Un/+c9/VLFiRat+27Zt8vJy7/9tFy5cqClTprgVaGrVqqVTp065vLcdLtbbqVOnVKECv14AiaAFoBScPXtWL730ku644w4tXry4wPjBgwdLoSv3BAUF6c9//rPLtldeeUVPPvmk3nrrLdWuXVuvvvqqNebr62trP2fPnlVeXp58fHzk5+dn63tdSmm/P1CWcOkQKEfS09PVtWtXBQYGKiAgQJ06ddLq1asL1GVmZmrgwIGqXbu2fH19VaNGDT388MP67bffJP3/pcqff/7Z5XUrVqyQw+HQihUrrG3bt29XQkKCIiIi5Ofnpxo1aqh79+7Kysoqss/ffvtNTqdTt956a6HjYWFhBd7zgw8+0NChQxUREaHKlSvr7rvv1p49e1xe99VXX+n+++9XzZo15evrq6ioKA0cOFCnTp1yqctfs/Trr7+qW7duCggIUGhoqJ599lnl5uYW2feleHt764033lDjxo01efJkl+/BhWu0cnJyNGrUKDVo0EB+fn665ppr1LZtW+vSY69evTRlyhRJsi5TOhwOSf+/Duu1117T66+/rnr16snX11c//vhjoWu08v33v/9VXFycKleurMjISI0ePVrGmALf6/OP7/nvl7/Pi/WWv+3CM13F+dnM/7n75ptvlJycrNDQUFWuXFn33nuvDh06dOkDAJRBnNECyonNmzerXbt2CgwM1HPPPaeKFSvq73//u2677TalpqYqOjpaknT8+HG1a9dOW7ZsUe/evdWyZUv99ttv+uyzz7R3715Vq1at2O955swZxcXFKTs7WwMGDFBERIR+/fVXzZ8/X5mZmQoKCir0dWFhYfL399fnn3+uAQMGKCQk5JLvNWbMGDkcDg0ePFgHDx7U66+/rtjYWK1fv17+/v6SpDlz5ujkyZPq16+frrnmGn377bd68803tXfvXs2ZM8dlf7m5uYqLi1N0dLRee+01LVmyRBMmTFC9evXUr1+/Yn8PLuTt7a0ePXpo2LBh+vrrrxUfH19o3ciRIzV27Fj95S9/0c033yyn06m1a9fq+++/1x133KHHHntM+/btU0pKiv79738Xuo8ZM2bo9OnTevTRR+Xr66uQkBDl5eUVWpubm6suXbqoTZs2GjdunBYtWqQRI0bo7NmzGj16tFtzLE5v5yvuz2a+AQMGqGrVqhoxYoR+/vlnvf7660pKStIHH3zgVp9AmWAAlHkzZswwksx3331XZE23bt2Mj4+P2blzp7Vt3759pkqVKqZ9+/bWtuHDhxtJ5pNPPimwj7y8PJf327Vrl8v48uXLjSSzfPlyY4wx6enpRpKZM2eO23PK76Ny5cqma9euZsyYMWbdunUF6vLf89prrzVOp9Pa/uGHHxpJZtKkSda2kydPFnj92LFjjcPhML/88ou1LTEx0Ugyo0ePdqm98cYbTatWrS7Ze4cOHcwNN9xQ5PjcuXML9FarVi2TmJhoPW/evLmJj4+/6Pv079/fFPaf6V27dhlJJjAw0Bw8eLDQsRkzZljb8uc7YMAAa1teXp6Jj483Pj4+5tChQ8aYgsf3YvssqjdjjJFkRowYYT0v7s9m/s9dbGys9bNojDEDBw403t7eJjMzs9D3A8oyLh0C5UBubq4WL16sbt26qW7dutb26tWr68EHH9TXX38tp9MpSfr444/VvHlz3XvvvQX2c/7ln+LIP2P15Zdf6uTJk269dtSoUZo1a5ZuvPFGffnll3rhhRfUqlUrtWzZUlu2bClQ//DDD6tKlSrW8z/+8Y+qXr26Fi5caG3LP7MlnVug/ttvv+mWW26RMUbp6ekF9vn444+7PG/Xrp3++9//ujWPwuTfSuHYsWNF1gQHB2vz5s3avn37Zb9PQkKCQkNDi12flJRkfe1wOJSUlKQzZ85oyZIll93Dpbjzs5nv0UcfdflZbNeunXJzc/XLL7/Y1idgF4IWUA4cOnRIJ0+eVMOGDQuMXX/99crLy7PWM+3cuVNNmjQpkfetU6eOkpOT9fbbb6tatWqKi4vTlClTLro+63w9evTQV199paNHj2rx4sV68MEHlZ6errvuukunT592qW3QoIHLc4fDofr167usI9u9e7d69eqlkJAQa91Vhw4dJKlAT35+fgVCStWqVXX06NHiTr9Ix48flySXYHih0aNHKzMzU9ddd52aNm2qQYMGaePGjW69T506dYpd6+Xl5RJ0JOm6666TpAJr8UqSOz+b+WrWrOnyvGrVqpJUIscGuNIIWgAKKOrMVmELxSdMmKCNGzdq6NChOnXqlJ588kndcMMN2rt3b7HfLzAwUHfccYfee+89JSYmaufOnVqzZo1bPefm5uqOO+7QggULNHjwYM2bN08pKSnWAu4L1y55e3u7tX93bNq0SZJUv379Imvat2+vnTt36p133lGTJk309ttvq2XLlnr77beL/T7nn8ErCe4cdzsVdWzMeQv3AU9B0ALKgdDQUFWqVEnbtm0rMLZ161Z5eXkpKipKklSvXj0rCBQl/wxCZmamy/aiLt00bdpUL774olauXKmvvvpKv/76q6ZNm3YZM5Fat24tSdq/f7/L9gsvsRljtGPHDtWuXVuS9MMPP+inn37ShAkTNHjwYN1zzz2KjY1VZGTkZfVxuXJzczVr1ixVqlRJbdu2vWhtSEiIHnnkEb3//vvas2ePmjVr5vJpPXcv5V5MXl5egcuiP/30kyRZ30N3jntxe3PnZxMojwhaQDng7e2tzp0769NPP3W5DHTgwAHNmjVLbdu2VWBgoKRz63o2bNiguXPnFthP/hmDevXqSZJWrlxpjeXm5mr69Oku9U6nU2fPnnXZ1rRpU3l5eSk7O7vIfk+ePKm0tLRCx7744gtJKnCp6V//+pfLmqePPvpI+/fvV9euXa3vwflzyP960qRJRfZR0nJzc/Xkk09qy5YtevLJJ63veWEOHz7s8jwgIED169d3+b5VrlxZUsHgc7kmT55sfW2M0eTJk1WxYkV16tRJ0rmbnXp7e7scd0l66623CuyruL2587MJlEfc3gHwIO+8806hf57mqaee0ssvv6yUlBS1bdtWTzzxhCpUqKC///3vys7O1rhx46zaQYMG6aOPPtL999+v3r17q1WrVjpy5Ig+++wzTZs2Tc2bN9cNN9ygNm3aaMiQITpy5IhCQkI0e/bsAqFq2bJlSkpK0v3336/rrrtOZ8+e1b///W95e3srISGhyHmcPHlSt9xyi9q0aaMuXbooKipKmZmZmjdvnr766it169ZNN954o8trQkJC1LZtWz3yyCM6cOCAXn/9ddWvX199+/aVJDVq1Ej16tXTs88+q19//VWBgYH6+OOPbVvXk5WVpf/85z/WfPLvDL9z5051795dL7300kVf37hxY912221q1aqVQkJCtHbtWn300UcuC9ZbtWolSXryyScVFxcnb29vde/e/bL69fPz06JFi5SYmKjo6Gh98cUXWrBggYYOHWqtVQsKCtL999+vN998Uw6HQ/Xq1dP8+fMLvYGsO70V92cTKJdK8ROPAIop/2PvRT327NljjDHm+++/N3FxcSYgIMBUqlTJdOzY0axatarA/g4fPmySkpLMtddea3x8fEyNGjVMYmKi+e2336yanTt3mtjYWOPr62vCw8PN0KFDTUpKisvH///73/+a3r17m3r16hk/Pz8TEhJiOnbsaJYsWXLR+eTk5Jh//OMfplu3bqZWrVrG19fXVKpUydx4441m/PjxJjs726rNv+XA+++/b4YMGWLCwsKMv7+/iY+Pd7llgzHG/PjjjyY2NtYEBASYatWqmb59+5oNGzYUeruDypUrF+hrxIgRRd6y4HwdOnRw+f4HBASYBg0amD//+c9m8eLFhb7mwts7vPzyy+bmm282wcHBxt/f3zRq1MiMGTPGnDlzxqo5e/asGTBggAkNDTUOh8PqLf92C+PHjy/wPkXd3qFy5cpm586dpnPnzqZSpUomPDzcjBgxwuTm5rq8/tChQyYhIcFUqlTJVK1a1Tz22GNm06ZNBfZZVG/GFLy9gzHF+9ks6jYmRd12AvAEDmNYXQig7FqxYoU6duyoOXPm6I9//GNptwMAbmGNFgAAgE0IWgAAADYhaAEAANiENVoAAAA24YwWAACATQhaAAAANuGGpSUkLy9P+/btU5UqVUr0z2YAAAD7GGN07NgxRUZGysur5M8/EbRKyL59+/h7XQAAeKg9e/aoRo0aJb5fglYJqVKliqRzB4q/2wUAgGdwOp2Kioqyfo+XNIJWCcm/XBgYGEjQAgDAw9i17IfF8AAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANikQmk3UN4ZY9x+jcPhsKETAABwpRG0roDuf18lFSc8GaPZj91if0MAAOCKIGhdCQ5Hsc5SuX/uCwAAlGWs0QIAALAJQQsAAMAmBC0AAACbELQAAABsQtACAACwCUELAADAJgQtAAAAmxC0AAAAbELQAgAAsAlBCwAAwCYELQAAAJsQtAAAAGxC0AIAALAJQQsAAMAmBC0AAACbELQAAABsQtACAACwCUELAADAJgQtAAAAmxC0AAAAbELQAgAAsAlBCwAAwCYELQAAAJsQtAAAAGxC0AIAALAJQQsAAMAmBC0AAACbELQAAABsQtACAACwCUELAADAJgQtAAAAmxC0AAAAbELQAgAAsAlBCwAAwCYELQAAAJsQtAAAAGxC0AIAALAJQQsAAMAmBC0AAACbELQAAABsQtACAACwCUELAADAJqUatMaOHaubbrpJVapUUVhYmLp166Zt27a51Jw+fVr9+/fXNddco4CAACUkJOjAgQMuNbt371Z8fLwqVaqksLAwDRo0SGfPnnWpWbFihVq2bClfX1/Vr19fM2fOLNDPlClTVLt2bfn5+Sk6Olrffvttic8ZAABcPUo1aKWmpqp///5avXq1UlJSlJOTo86dO+vEiRNWzcCBA/X5559rzpw5Sk1N1b59+3TfffdZ47m5uYqPj9eZM2e0atUqvfvuu5o5c6aGDx9u1ezatUvx8fHq2LGj1q9fr6efflp/+ctf9OWXX1o1H3zwgZKTkzVixAh9//33at68ueLi4nTw4MEr880AAADljsMYY0q7iXyHDh1SWFiYUlNT1b59e2VlZSk0NFSzZs3SH//4R0nS1q1bdf311ystLU1t2rTRF198oT/84Q/at2+fwsPDJUnTpk3T4MGDdejQIfn4+Gjw4MFasGCBNm3aZL1X9+7dlZmZqUWLFkmSoqOjddNNN2ny5MmSpLy8PEVFRWnAgAF6/vnnL9m70+lUUFCQsrKyFBgYaG03xqj79DQ5HI5L7sMYo9mPxhSrFgAA/H5F/f4uKWVqjVZWVpYkKSQkRJK0bt065eTkKDY21qpp1KiRatasqbS0NElSWlqamjZtaoUsSYqLi5PT6dTmzZutmvP3kV+Tv48zZ85o3bp1LjVeXl6KjY21agAAANxVobQbyJeXl6enn35at956q5o0aSJJysjIkI+Pj4KDg11qw8PDlZGRYdWcH7Lyx/PHLlbjdDp16tQpHT16VLm5uYXWbN26tdB+s7OzlZ2dbT13Op1uzhgAAJR3ZeaMVv/+/bVp0ybNnj27tFsplrFjxyooKMh6REVFlXZLAACgjCkTQSspKUnz58/X8uXLVaNGDWt7RESEzpw5o8zMTJf6AwcOKCIiwqq58FOI+c8vVRMYGCh/f39Vq1ZN3t7ehdbk7+NCQ4YMUVZWlvXYs2eP+xMHAADlWqkGLWOMkpKSNHfuXC1btkx16tRxGW/VqpUqVqyopUuXWtu2bdum3bt3KyYmRpIUExOjH374weXTgSkpKQoMDFTjxo2tmvP3kV+Tvw8fHx+1atXKpSYvL09Lly61ai7k6+urwMBAlwcAAMD5SnWNVv/+/TVr1ix9+umnqlKlirWmKigoSP7+/goKClKfPn2UnJyskJAQBQYGasCAAYqJiVGbNm0kSZ07d1bjxo310EMPady4ccrIyNCLL76o/v37y9fXV5L0+OOPa/LkyXruuefUu3dvLVu2TB9++KEWLFhg9ZKcnKzExES1bt1aN998s15//XWdOHFCjzzyyJX/xgAAgHKhVIPW1KlTJUm33Xaby/YZM2aoV69ekqSJEyfKy8tLCQkJys7OVlxcnN566y2r1tvbW/Pnz1e/fv0UExOjypUrKzExUaNHj7Zq6tSpowULFmjgwIGaNGmSatSoobfffltxcXFWzQMPPKBDhw5p+PDhysjIUIsWLbRo0aICC+QBAACKq0zdR8uTcR8tAAA8z1V1Hy0AAIDyhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0qlHYD+H/GGBlj3HqNw+GwqRsAAPB7EbTKmB7T06TihCdjNPuxW+xvCAAAXDaCVlnjcBTrLJV7570AAEBpYI0WAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANikVIPWypUrdddddykyMlIOh0Pz5s1zGe/Vq5ccDofLo0uXLi41R44cUc+ePRUYGKjg4GD16dNHx48fd6nZuHGj2rVrJz8/P0VFRWncuHEFepkzZ44aNWokPz8/NW3aVAsXLizx+QIAgKtLqQatEydOqHnz5poyZUqRNV26dNH+/futx/vvv+8y3rNnT23evFkpKSmaP3++Vq5cqUcffdQadzqd6ty5s2rVqqV169Zp/PjxGjlypKZPn27VrFq1Sj169FCfPn2Unp6ubt26qVu3btq0aVPJTxoAAFw1HMYYU9pNSJLD4dDcuXPVrVs3a1uvXr2UmZlZ4ExXvi1btqhx48b67rvv1Lp1a0nSokWLdOedd2rv3r2KjIzU1KlT9cILLygjI0M+Pj6SpOeff17z5s3T1q1bJUkPPPCATpw4ofnz51v7btOmjVq0aKFp06YVq3+n06mgoCBlZWUpMDDQ2m6MUffpaXI4HJfcR15ennXm7lKMMZr9aEyxagEAQOGK+v1dUsr8Gq0VK1YoLCxMDRs2VL9+/XT48GFrLC0tTcHBwVbIkqTY2Fh5eXlpzZo1Vk379u2tkCVJcXFx2rZtm44ePWrVxMbGurxvXFyc0tLSiuwrOztbTqfT5QEAAHC+Mh20unTpon/9619aunSpXn31VaWmpqpr167Kzc2VJGVkZCgsLMzlNRUqVFBISIgyMjKsmvDwcJea/OeXqskfL8zYsWMVFBRkPaKion7fZAEAQLlTobQbuJju3btbXzdt2lTNmjVTvXr1tGLFCnXq1KkUO5OGDBmi5ORk67nT6SRsAQAAF26f0br99tuVmZlZYLvT6dTtt99eEj0VqW7duqpWrZp27NghSYqIiNDBgwddas6ePasjR44oIiLCqjlw4IBLTf7zS9XkjxfG19dXgYGBLg8AAIDzuR20VqxYoTNnzhTYfvr0aX311Vcl0lRR9u7dq8OHD6t69eqSpJiYGGVmZmrdunVWzbJly5SXl6fo6GirZuXKlcrJybFqUlJS1LBhQ1WtWtWqWbp0qct7paSkKCYmxtb5AACA8q3Ylw43btxoff3jjz+6rF/Kzc3VokWLdO2117r15sePH7fOTknSrl27tH79eoWEhCgkJESjRo1SQkKCIiIitHPnTj333HOqX7++4uLiJEnXX3+9unTpor59+2ratGnKyclRUlKSunfvrsjISEnSgw8+qFGjRqlPnz4aPHiwNm3apEmTJmnixInW+z711FPq0KGDJkyYoPj4eM2ePVtr1651uQUEAACAu4odtFq0aGHdeqCwS4T+/v5688033XrztWvXqmPHjtbz/DVPiYmJmjp1qjZu3Kh3331XmZmZioyMVOfOnfXSSy/J19fXes17772npKQkderUSV5eXkpISNAbb7xhjQcFBWnx4sXq37+/WrVqpWrVqmn48OEu99q65ZZbNGvWLL344osaOnSoGjRooHnz5qlJkyZuzQcAAOB8xb6P1i+//CJjjOrWratvv/1WoaGh1piPj4/CwsLk7e1tW6NlHffRAgDA89h9H61in9GqVauWpHNhAAAAAJd2Wbd32L59u5YvX66DBw8WCF7Dhw8vkcYAAAA8ndtB6x//+If69eunatWqKSIiwuXSlcPhIGgBAAD8j9tB6+WXX9aYMWM0ePBgO/oBAAAoN9y+j9bRo0d1//3329ELAABAueJ20Lr//vu1ePFiO3oBAAAoV9y+dFi/fn0NGzZMq1evVtOmTVWxYkWX8SeffLLEmgMAAPBkbget6dOnKyAgQKmpqUpNTXUZczgcBC0AAID/cTto7dq1y44+AAAAyh2312gBAACgeNw+o9W7d++Ljr/zzjuX3QwAAEB54nbQOnr0qMvznJwcbdq0SZmZmYX+sWkAAICrldtBa+7cuQW25eXlqV+/fqpXr16JNAUAAFAelMgaLS8vLyUnJ2vixIklsTsAAIByocQWw+/cuVNnz54tqd0BAAB4PLcvHSYnJ7s8N8Zo//79WrBggRITE0usMQAAAE/ndtBKT093ee7l5aXQ0FBNmDDhkp9IBAAAuJq4HbSWL19uRx8AAADljttBK9+hQ4e0bds2SVLDhg0VGhpaYk0BAACUB24vhj9x4oR69+6t6tWrq3379mrfvr0iIyPVp08fnTx50o4eAQAAPJLbQSs5OVmpqan6/PPPlZmZqczMTH366adKTU3VM888Y0ePAAAAHsntS4cff/yxPvroI912223WtjvvvFP+/v7605/+pKlTp5ZkfwAAAB7L7TNaJ0+eVHh4eIHtYWFhXDoEAAA4j9tBKyYmRiNGjNDp06etbadOndKoUaMUExNTos0BAAB4MrcvHU6aNElxcXGqUaOGmjdvLknasGGD/Pz89OWXX5Z4gwAAAJ7K7aDVpEkTbd++Xe+99562bt0qSerRo4d69uwpf3//Em8QAADAU13WfbQqVaqkvn37lnQvAAAA5Uqx12itW7dOHTt2lNPpLDCWlZWljh07asOGDSXaHAAAgCcrdtCaMGGCbr/9dgUGBhYYCwoK0h133KHx48eXaHMAAACerNhBa82aNbrnnnuKHL/rrru0atWqEmkKAACgPCh20Pr1119VpUqVIscDAgK0f//+EmkKAACgPCh20AoNDbX+iHRhtm7dqmrVqpVIUwAAAOVBsYNWbGysxowZU+iYMUZjxoxRbGxsiTUGAADg6Yp9e4cXX3xRrVq1UnR0tJ555hk1bNhQ0rkzWRMmTNBPP/2kmTNn2tUnAACAxyl20KpXr56WLFmiXr16qXv37nI4HJLOnc1q3LixUlJSVL9+fdsaBQAA8DRu3bC0devW2rRpk9avX6/t27fLGKPrrrtOLVq0sKk9AAAAz3VZd4Zv0aIF4QoAAOASir0YHgAAAO4haAEAANiEoAUAAGATghYAAIBN3A5aixYt0tdff209nzJlilq0aKEHH3xQR48eLdHmAAAAPJnbQWvQoEFyOp2SpB9++EHPPPOM7rzzTu3atUvJyckl3iAAAICncvv2Drt27VLjxo0lSR9//LH+8Ic/6K9//au+//573XnnnSXeIAAAgKdy+4yWj4+PTp48KUlasmSJOnfuLEkKCQmxznQBAADgMs5otW3bVsnJybr11lv17bff6oMPPpAk/fTTT6pRo0aJNwgAAOCp3D6jNXnyZFWoUEEfffSRpk6dqmuvvVaS9MUXX6hLly4l3iAAAICncvuMVs2aNTV//vwC2ydOnFgiDQEAAJQXbget3bt3X3S8Zs2al90MAABAeeJ20Kpdu7YcDkeR47m5ub+rIQAAgPLC7aCVnp7u8jwnJ0fp6en629/+pjFjxpRYYwAAAJ7O7aDVvHnzAttat26tyMhIjR8/Xvfdd1+JNAYAAODpSuxvHTZs2FDfffddSe0OAADA47l9RuvCm5IaY7R//36NHDlSDRo0KLHGAAAAPJ3bQSs4OLjAYnhjjKKiojR79uwSawwAAMDTuR20li1b5hK0vLy8FBoaqvr166tCBbd3BwAAUG65nYxuu+02G9oAAAAof9xeDD927Fi98847Bba/8847evXVV0ukKQAAgPLA7aD197//XY0aNSqw/YYbbtC0adNKpCkAAIDywO2glZGRoerVqxfYHhoaqv3795dIUwAAAOWB20ErKipK33zzTYHt33zzjSIjI0ukKQAAgPLA7cXwffv21dNPP62cnBzdfvvtkqSlS5fqueee0zPPPFPiDQIAAHgqt4PWoEGDdPjwYT3xxBM6c+aMJMnPz0+DBw/WkCFDSrxBAAAAT+V20HI4HHr11Vc1bNgwbdmyRf7+/mrQoIF8fX3t6A8AAMBjXfYdRgMCAqxF8YQsAACAgtxeDJ+Xl6fRo0crKChItWrVUq1atRQcHKyXXnpJeXl5bu1r5cqVuuuuuxQZGSmHw6F58+a5jBtjNHz4cFWvXl3+/v6KjY3V9u3bXWqOHDminj17KjAwUMHBwerTp4+OHz/uUrNx40a1a9dOfn5+ioqK0rhx4wr0MmfOHDVq1Eh+fn5q2rSpFi5c6NZcAAAALuR20HrhhRc0efJkvfLKK0pPT1d6err++te/6s0339SwYcPc2teJEyfUvHlzTZkypdDxcePG6Y033tC0adO0Zs0aVa5cWXFxcTp9+rRV07NnT23evFkpKSmaP3++Vq5cqUcffdQadzqd6ty5s2rVqqV169Zp/PjxGjlypKZPn27VrFq1Sj169FCfPn2Unp6ubt26qVu3btq0aZOb3x0AAID/5zDGGHdeEBkZqWnTpunuu+922f7pp5/qiSee0K+//np5jTgcmjt3rrp16ybp3NmsyMhIPfPMM3r22WclSVlZWQoPD9fMmTPVvXt3bdmyRY0bN9Z3332n1q1bS5IWLVqkO++8U3v37lVkZKSmTp2qF154QRkZGfLx8ZEkPf/885o3b562bt0qSXrggQd04sQJzZ8/3+qnTZs2atGiRbFvwup0OhUUFKSsrCwFBgZa240x6j49rcAf4i5MXl6eHA5HsWqNMZr9aEyxagEAQOGK+v1dUtw+o3XkyJFC7wzfqFEjHTlypESakqRdu3YpIyNDsbGx1ragoCBFR0crLS1NkpSWlqbg4GArZElSbGysvLy8tGbNGqumffv2VsiSpLi4OG3btk1Hjx61as5/n/ya/PcpTHZ2tpxOp8sDAADgfG4HrebNm2vy5MkFtk+ePFnNmzcvkaakc3egl6Tw8HCX7eHh4dZYRkaGwsLCXMYrVKigkJAQl5rC9nH+exRVkz9emLFjxyooKMh6REVFuTtFAABQzrn9qcNx48YpPj5eS5YsUUxMjKRzZ4T27NlzVS0gHzJkiJKTk63nTqeTsAUAAFy4fUarQ4cO+umnn3TvvfcqMzNTmZmZuu+++7Rt2za1a9euxBqLiIiQJB04cMBl+4EDB6yxiIgIHTx40GX87NmzOnLkiEtNYfs4/z2KqskfL4yvr68CAwNdHgAAAOdzO2hJ5xbEjxkzRh9//LE+/vhjvfzyy8rLy3P5tN/vVadOHUVERGjp0qXWNqfTqTVr1lhn0mJiYpSZmal169ZZNcuWLVNeXp6io6OtmpUrVyonJ8eqSUlJUcOGDVW1alWr5vz3ya/Jfx8AAIDLcVlBqzCHDx/WP//5T7dec/z4ca1fv17r16+XdG4B/Pr167V79245HA49/fTTevnll/XZZ5/phx9+0MMPP6zIyEjrk4nXX3+9unTpor59++rbb7/VN998o6SkJHXv3t36A9cPPvigfHx81KdPH23evFkffPCBJk2a5HLZ76mnntKiRYs0YcIEbd26VSNHjtTatWuVlJRUIt8bAABwdbrsO8OXhLVr16pjx47W8/zwk5iYqJkzZ+q5557TiRMn9OijjyozM1Nt27bVokWL5OfnZ73mvffeU1JSkjp16iQvLy8lJCTojTfesMaDgoK0ePFi9e/fX61atVK1atU0fPhwl7Nvt9xyi2bNmqUXX3xRQ4cOVYMGDTRv3jw1adLkCnwXAABAeeX2fbSKsmHDBrVs2VK5ubklsTuPw320AADwPGXuPloAAAAonmJfOrzvvvsuOp6Zmfl7ewEAAChXih20goKCLjn+8MMP/+6GAAAAyotiB60ZM2bY2QcAAEC5wxotAAAAmxC0AAAAbELQAgAAsAlBCwAAwCYELQAAAJsQtAAAAGxC0AIAALAJQQsAAMAmBC0AAACbFPvO8ChbjDEyxrj1GofDYVM3AACgMAQtD9ZjeppUnPBkjGY/dov9DQEAABcELU/mcBTrLJV7570AAEBJYY0WAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANqlQ2g3AfsYYGWPceo3D4bCpGwAArh4EratEj+lpUnHCkzGa/dgt9jcEAMBVgKB1tXA4inWWyr3zXgAA4GJYowUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYBOCFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2KRMB62RI0fK4XC4PBo1amSNnz59Wv3799c111yjgIAAJSQk6MCBAy772L17t+Lj41WpUiWFhYVp0KBBOnv2rEvNihUr1LJlS/n6+qp+/fqaOXPmlZgeAAAo58p00JKkG264Qfv377ceX3/9tTU2cOBAff7555ozZ45SU1O1b98+3XfffdZ4bm6u4uPjdebMGa1atUrvvvuuZs6cqeHDh1s1u3btUnx8vDp27Kj169fr6aef1l/+8hd9+eWXV3SeAACg/KlQ2g1cSoUKFRQREVFge1ZWlv75z39q1qxZuv322yVJM2bM0PXXX6/Vq1erTZs2Wrx4sX788UctWbJE4eHhatGihV566SUNHjxYI0eOlI+Pj6ZNm6Y6depowoQJkqTrr79eX3/9tSZOnKi4uLgrOlcAAFC+lPkzWtu3b1dkZKTq1q2rnj17avfu3ZKkdevWKScnR7GxsVZto0aNVLNmTaWlpUmS0tLS1LRpU4WHh1s1cXFxcjqd2rx5s1Vz/j7ya/L3UZTs7Gw5nU6XBwAAwPnKdNCKjo7WzJkztWjRIk2dOlW7du1Su3btdOzYMWVkZMjHx0fBwcEurwkPD1dGRoYkKSMjwyVk5Y/nj12sxul06tSpU0X2NnbsWAUFBVmPqKio3ztdAABQzpTpS4ddu3a1vm7WrJmio6NVq1Ytffjhh/L39y/FzqQhQ4YoOTnZeu50OglbAADARZk+o3Wh4OBgXXfdddqxY4ciIiJ05swZZWZmutQcOHDAWtMVERFR4FOI+c8vVRMYGHjRMOfr66vAwECXR3lgjHH7AQAAClemz2hd6Pjx49q5c6ceeughtWrVShUrVtTSpUuVkJAgSdq2bZt2796tmJgYSVJMTIzGjBmjgwcPKiwsTJKUkpKiwMBANW7c2KpZuHChy/ukpKRY+7ga9ZieJjkcly40RrMfu8X+hgAA8FBl+ozWs88+q9TUVP38889atWqV7r33Xnl7e6tHjx4KCgpSnz59lJycrOXLl2vdunV65JFHFBMTozZt2kiSOnfurMaNG+uhhx7Shg0b9OWXX+rFF19U//795evrK0l6/PHH9d///lfPPfectm7dqrfeeksffvihBg4cWJpTL10X3LusqEexwhgAAFexMn1Ga+/everRo4cOHz6s0NBQtW3bVqtXr1ZoaKgkaeLEifLy8lJCQoKys7MVFxent956y3q9t7e35s+fr379+ikmJkaVK1dWYmKiRo8ebdXUqVNHCxYs0MCBAzVp0iTVqFFDb7/9Nrd2AAAAv5vDsMimRDidTgUFBSkrK8tlvZYxRt2np507A3QJeXl5/3+2yANqjTGa/WhMsWoBACiLivr9XVLK9KVDAAAAT0bQAgAAsAlBCwAAwCYELQAAAJsQtAAAAGxC0AIAALAJQQsAAMAmBC0AAACbELQAAABsQtACAACwCUELAADAJgQtAAAAmxC0AAAAbELQAgAAsAlBCwAAwCYELQAAAJsQtAAAAGxC0AIAALAJQQsAAMAmBC0AAACbELQAAABsQtACAACwSYXSbgCeyxgjY4zbr3M4HDZ0AwBA2UPQwu/SY3qaVNzgZIxmP3aLvQ0BAFCGELTw+zgcxT5D5f65LwAAPBtrtAAAAGxC0AIAALAJQQsAAMAmBC0AAACbELQAAABsQtACAACwCUELAADAJgQtAAAAmxC0AAAAbELQAgAAsAlBCwAAwCb8rUNcMcYYGePeXzws7t9RBACgLCJo4YrqMT1NKk54MkazH7vF/oYAALARQQtXlsNRrLNU7p33AgCgbGKNFgAAgE0IWgAAADYhaAEAANiEoAUAAGATghYAAIBN+NQhyiTuuQUAKA8IWiizuOcWAMDTEbRQdnHPLQCAh2ONFgAAgE0IWgAAADYhaAEAANiENVrweHxCEQBQVhG0UC7wCUUAQFlE0EL5wCcUAQBlEGu0AAAAbMIZLVxVWM8FALiSCFq46rCeCwBwpRC0cPVhPRcA4AohaAFF4DIjAOD3ImgBF1Hcy4wmL8/ty4yEMgAo/whawMUU9zKjw8HaLwBAAQQtoKQUM5TlXcYlyXO75wwYAHgaghZQCop99ktclgQAT0bQAkpDMc9+Se5dliSUAUDZQtACPIENa8UIZQBgP4LWBaZMmaLx48crIyNDzZs315tvvqmbb765tNsCiq+UQ1n++rNi9eBG7fkIfAA8BUHrPB988IGSk5M1bdo0RUdH6/XXX1dcXJy2bdumsLCw0m4PKHk2hTKHw1Hitfn1ZSXwEfYAFIfDXM7Hn8qp6Oho3XTTTZo8ebIkKS8vT1FRURowYICef/75i77W6XQqKChIWVlZCgwMtLYbY9R9elrxPo32v1865bG2rPThabVlpY+yUGvVS6Ua+PLDXkkHODuDYVnooywEX3d/3Xna/HB5ivr9XVI4o/U/Z86c0bp16zRkyBBrm5eXl2JjY5WWllaKnQFw4cZZOLtqS/vs3uWcCSztPuwOvsVhjHH7E7+eNj+p9INhWejDndq8vLxL1vweBK3/+e2335Sbm6vw8HCX7eHh4dq6dWuB+uzsbGVnZ1vPs7KyJJ1Lxuczxijn5PFS/4da2rVlpQ9Pqy0rfZSF2rLSh6fVlpU+7KxNmJQih4rxC9Xkycvh5XHfN3fm53A4SrW2rPThTm32qWOS3D/jWVwErcs0duxYjRo1qsD2qKioUugGAAD8HocPH1ZQUFCJ75eg9T/VqlWTt7e3Dhw44LL9wIEDioiIKFA/ZMgQJScnW8/z8vJ05MgRXXPNNR59Ld3pdCoqKkp79uyx5Vp1aWN+nq08z688z01ifp6sPM9NOndFqmbNmgoJCbFl/wSt//Hx8VGrVq20dOlSdevWTdK58LR06VIlJSUVqPf19ZWvr6/LtuDg4CvQ6ZURGBhYLv9B5WN+nq08z688z01ifp6sPM9NOrcu2w4ErfMkJycrMTFRrVu31s0336zXX39dJ06c0COPPFLarQEAAA9E0DrPAw88oEOHDmn48OHKyMhQixYttGjRogIL5AEAAIqDoHWBpKSkQi8VXi18fX01YsSIApdFywvm59nK8/zK89wk5ufJyvPcJPvnxw1LAQAAbGLPyi8AAAAQtAAAAOxC0AIAALAJQQsAAMAmBK2rxMqVK3XXXXcpMjJSDodD8+bNcxk3xmj48OGqXr26/P39FRsbq+3bt7vUHDlyRD179lRgYKCCg4PVp08fHT9+/ArOonCXmluvXr3O/c2r8x5dunRxqSmrc5PO/bmnm266SVWqVFFYWJi6deumbdu2udScPn1a/fv31zXXXKOAgAAlJCQU+CsHu3fvVnx8vCpVqqSwsDANGjRIZ8+evZJTKaA4c7vtttsKHL/HH3/cpaYszk2Spk6dqmbNmlk3eoyJidEXX3xhjXvqcct3qfl58rG70CuvvCKHw6Gnn37a2ubpx+98hc3Pk4/fyJEjC/TeqFEja/yKHjuDq8LChQvNCy+8YD755BMjycydO9dl/JVXXjFBQUFm3rx5ZsOGDebuu+82derUMadOnbJqunTpYpo3b25Wr15tvvrqK1O/fn3To0ePKzyTgi41t8TERNOlSxezf/9+63HkyBGXmrI6N2OMiYuLMzNmzDCbNm0y69evN3feeaepWbOmOX78uFXz+OOPm6ioKLN06VKzdu1a06ZNG3PLLbdY42fPnjVNmjQxsbGxJj093SxcuNBUq1bNDBkypDSmZCnO3Dp06GD69u3rcvyysrKs8bI6N2OM+eyzz8yCBQvMTz/9ZLZt22aGDh1qKlasaDZt2mSM8dzjlu9S8/PkY3e+b7/91tSuXds0a9bMPPXUU9Z2Tz9++YqanycfvxEjRpgbbrjBpfdDhw5Z41fy2BG0rkIXhpG8vDwTERFhxo8fb23LzMw0vr6+5v333zfGGPPjjz8aSea7776zar744gvjcDjMr7/+esV6v5SigtY999xT5Gs8ZW75Dh48aCSZ1NRUY8y5Y1WxYkUzZ84cq2bLli1GkklLSzPGnAujXl5eJiMjw6qZOnWqCQwMNNnZ2Vd2Ahdx4dyMOfcf+/P/438hT5lbvqpVq5q33367XB238+XPz5jyceyOHTtmGjRoYFJSUlzmU16OX1HzM8azj9+IESNM8+bNCx270seOS4fQrl27lJGRodjYWGtbUFCQoqOjlZaWJklKS0tTcHCwWrdubdXExsbKy8tLa9asueI9u2vFihUKCwtTw4YN1a9fPx0+fNga87S5ZWVlSZL1B1DXrVunnJwcl+PXqFEj1axZ0+X4NW3a1OWvHMTFxcnpdGrz5s1XsPuLu3Bu+d577z1Vq1ZNTZo00ZAhQ3Ty5ElrzFPmlpubq9mzZ+vEiROKiYkpV8dNKji/fJ5+7Pr376/4+HiX4ySVn393Rc0vnycfv+3btysyMlJ169ZVz549tXv3bklX/thxZ3goIyNDkgr8qaHw8HBrLCMjQ2FhYS7jFSpUUEhIiFVTVnXp0kX33Xef6tSpo507d2ro0KHq2rWr0tLS5O3t7VFzy8vL09NPP61bb71VTZo0kXTu2Pj4+BT4o+YXHr/Cjm/+WFlQ2Nwk6cEHH1StWrUUGRmpjRs3avDgwdq2bZs++eQTSWV/bj/88INiYmJ0+vRpBQQEaO7cuWrcuLHWr19fLo5bUfOTPP/YzZ49W99//72+++67AmPl4d/dxeYnefbxi46O1syZM9WwYUPt379fo0aNUrt27bRp06YrfuwIWij3unfvbn3dtGlTNWvWTPXq1dOKFSvUqVOnUuzMff3799emTZv09ddfl3YrJa6ouT366KPW102bNlX16tXVqVMn7dy5U/Xq1bvSbbqtYcOGWr9+vbKysvTRRx8pMTFRqamppd1WiSlqfo0bN/boY7dnzx499dRTSklJkZ+fX2m3U+KKMz9PPn5du3a1vm7WrJmio6NVq1Ytffjhh/L397+ivXDpEIqIiJCkAp+4OHDggDUWERGhgwcPuoyfPXtWR44csWo8Rd26dVWtWjXt2LFDkufMLSkpSfPnz9fy5ctVo0YNa3tERITOnDmjzMxMl/oLj19hxzd/rLQVNbfCREdHS5LL8SvLc/Px8VH9+vXVqlUrjR07Vs2bN9ekSZPKxXGTip5fYTzp2K1bt04HDx5Uy5YtVaFCBVWoUEGpqal64403VKFCBYWHh3v08bvU/HJzcwu8xpOO34WCg4N13XXXaceOHVf83x5BC6pTp44iIiK0dOlSa5vT6dSaNWustRYxMTHKzMzUunXrrJply5YpLy/P+sfnKfbu3avDhw+revXqksr+3IwxSkpK0ty5c7Vs2TLVqVPHZbxVq1aqWLGiy/Hbtm2bdu/e7XL8fvjhB5dAmZKSosDAQOsyT2m41NwKs379eklyOX5lcW5FycvLU3Z2tkcft4vJn19hPOnYderUST/88IPWr19vPVq3bq2ePXtaX3vy8bvU/Ly9vQu8xpOO34WOHz+unTt3qnr16lf+3567K/nhmY4dO2bS09NNenq6kWT+9re/mfT0dPPLL78YY87d3iE4ONh8+umnZuPGjeaee+4p9PYON954o1mzZo35+uuvTYMGDcrELRAuNrdjx46ZZ5991qSlpZldu3aZJUuWmJYtW5oGDRqY06dPW/soq3Mzxph+/fqZoKAgs2LFCpePKp88edKqefzxx03NmjXNsmXLzNq1a01MTIyJiYmxxvM/qty5c2ezfv16s2jRIhMaGlrqH8O+1Nx27NhhRo8ebdauXWt27dplPv30U1O3bl3Tvn17ax9ldW7GGPP888+b1NRUs2vXLrNx40bz/PPPG4fDYRYvXmyM8dzjlu9i8/P0Y1eYCz+F5+nH70Lnz8/Tj98zzzxjVqxYYXbt2mW++eYbExsba6pVq2YOHjxojLmyx46gdZVYvny5kVTgkZiYaIw5d4uHYcOGmfDwcOPr62s6depktm3b5rKPw4cPmx49epiAgAATGBhoHnnkEXPs2LFSmI2ri83t5MmTpnPnziY0NNRUrFjR1KpVy/Tt29flI7vGlN25GWMKnZskM2PGDKvm1KlT5oknnjBVq1Y1lSpVMvfee6/Zv3+/y35+/vln07VrV+Pv72+qVatmnnnmGZOTk3OFZ+PqUnPbvXu3ad++vQkJCTG+vr6mfv36ZtCgQS738jGmbM7NGGN69+5tatWqZXx8fExoaKjp1KmTFbKM8dzjlu9i8/P0Y1eYC4OWpx+/C50/P08/fg888ICpXr268fHxMddee6154IEHzI4dO6zxK3nsHMYY4945MAAAABQHa7QAAABsQtACAACwCUELAADAJgQtAAAAmxC0AAAAbELQAgAAsAlBCwAAwCYELQAAAJsQtABcNQ4dOqR+/fqpZs2a8vX1VUREhOLi4vTNN9+UdmsAyqkKpd0AAFwpCQkJOnPmjN59913VrVtXBw4c0NKlS3X48OHSbg1AOcUZLQBXhczMTH311Vd69dVX1bFjR9WqVUs333yzhgwZorvvvluS5HA4NHXqVHXt2lX+/v6qW7euPvroI5f9DB48WNddd50qVaqkunXratiwYcrJybHGR44cqRYtWujf//63ateuraCgIHXv3l3Hjh27ovMFUDYQtABcFQICAhQQEKB58+YpOzu7yLphw4YpISFBGzZsUM+ePdW9e3dt2bLFGq9SpYpmzpypH3/8UZMmTdI//vEPTZw40WUfO3fu1Lx58zR//nzNnz9fqampeuWVV2ybG4Cyiz8qDeCq8fHHH6tv3746deqUWrZsqQ4dOqh79+5q1qyZpHNntB5//HFNnTrVek2bNm3UsmVLvfXWW4Xu87XXXtPs2bO1du1aSefOaI0fP14ZGRmqUqWKJOm5557TypUrtXr1aptnCKCs4YwWgKtGQkKC9u3bp88++0xdunTRihUr1LJlS82cOdOqiYmJcXlNTEyMyxmtDz74QLfeeqsiIiIUEBCgF198Ubt373Z5Te3ata2QJUnVq1fXwYMH7ZkUgDKNoAXgquLn56c77rhDw4YN06pVq9SrVy+NGDGiWK9NS0tTz549deedd2r+/PlKT0/XCy+8oDNnzrjUVaxY0eW5w+FQXl5eic0BgOcgaAG4qjVu3FgnTpywnl94eW/16tW6/vrrJUmrVq1SrVq19MILL6h169Zq0KCBfvnllyvaLwDPwu0dAFwVDh8+rPvvv1+9e/dWs2bNVKVKFa1du1bjxo3TPffcY9XNmTNHrVu3Vtu2bfXee+/p22+/1T//+U9JUoMGDbR7927Nnj1bN910kxYsWKC5c+eW1pQAeACCFoCrQkBAgKKjozVx4kTt3LlTOTk5ioqKUt++fTV06FCrbtSoUZo9e7aeeOIJVa9eXe+//74aN24sSbr77rs1cOBAJSUlKTs7W/Hx8Ro2bJhGjhxZSrMCUNbxqUMA+B+Hw6G5c+eqW7dupd0KgHKCNVoAAAA2IWgBAADYhDVaAPA/rKQAUNI4owUAAGATghYAAIBNCFoAAAA2IWgBAADYhKAFAABgE4IWAACATQhaAAAANiFoAQAA2ISgBQAAYJP/Az5nWrgqJP/JAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>77213.0</td>\n",
       "      <td>122.810149</td>\n",
       "      <td>382.31894</td>\n",
       "      <td>60.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>48491.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     count        mean        std   min   25%   50%    75%      max\n",
       "0  77213.0  122.810149  382.31894  60.0  67.0  78.0  104.0  48491.0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "p = sb.histplot(locus_span, binwidth=10)\n",
    "p.set(title=\"Locus Span Distribution\", \n",
    "      xlabel=\"Span\", \n",
    "      ylabel=\"Locus Count\",\n",
    "      xlim=(55,500)) # Zoom in to make the plot more attractive\n",
    "plt.show()\n",
    "\n",
    "locus_span.describe().to_frame().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2365f32d-05f1-4a4a-b573-7b56db496534",
   "metadata": {},
   "source": [
    "## The Allele table\n",
    "\n",
    "Every locus can have multiple observed alleles. There's always the reference allele, and then some number of alternate alleles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c76e61c4-bcd9-422b-b625-f047cd23aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusID</th>\n",
       "      <th>allele_number</th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>165</td>\n",
       "      <td>TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTGTGTCTCCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTCTGTCTCCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>163</td>\n",
       "      <td>TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTGTGTCTCCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>167</td>\n",
       "      <td>TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTGTGTCTCCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>GTGTATATGCTCCACGATGCCTGTGAATATACACACACACCACATC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   LocusID  allele_number  allele_length  \\\n",
       "0        0              0            165   \n",
       "1        0              1            139   \n",
       "2        0              2            163   \n",
       "3        0              3            167   \n",
       "4        1              0             70   \n",
       "\n",
       "                                            sequence  \n",
       "0  TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTGTGTCTCCCTC...  \n",
       "1  TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTCTGTCTCCCTC...  \n",
       "2  TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTGTGTCTCCCTC...  \n",
       "3  TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTGTGTCTCCCTC...  \n",
       "4  GTGTATATGCTCCACGATGCCTGTGAATATACACACACACCACATC...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = data[\"allele\"] # Again, just making a convience variable\n",
    "allele.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45187356-55cf-40f6-a7f0-a2eb4002ef06",
   "metadata": {},
   "source": [
    "There are four columns in the allele table.\n",
    "\n",
    "- LocusID : The primary key that ties the allele to its locus\n",
    "- allele_number : The identifier for this allele at the locus. Note that `allele_number == 0` is always the reference allele\n",
    "- allele_length : The length of the allele\n",
    "- sequence : The sequence of the allele\n",
    "\n",
    "We can again use pandas to get some basic summaries of our our alleles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2bcd742b-21aa-44c7-a898-1a46ad0621d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What is the distribution of allele lengths?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>213156.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>143.178527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>398.290971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>68.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>79.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>108.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>48491.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       allele_length\n",
       "count  213156.000000\n",
       "mean      143.178527\n",
       "std       398.290971\n",
       "min         9.000000\n",
       "25%        68.000000\n",
       "50%        79.000000\n",
       "75%       108.000000\n",
       "max     48491.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "How many alleles are there per-locus?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>77213.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>2.760623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.185621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            LocusID\n",
       "count  77213.000000\n",
       "mean       2.760623\n",
       "std        1.185621\n",
       "min        1.000000\n",
       "25%        2.000000\n",
       "50%        3.000000\n",
       "75%        3.000000\n",
       "max        7.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display # Helps display pretty Tables\n",
    "\n",
    "print(\"What is the distribution of allele lengths?\")\n",
    "display(allele['allele_length'].describe().to_frame())\n",
    "\n",
    "print(\"\\n\\nHow many alleles are there per-locus?\")\n",
    "alleles_per_locus = allele['LocusID'].value_counts()\n",
    "display(alleles_per_locus.describe().to_frame())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e6b16df-a7b3-4c02-81d3-633d6b90ec97",
   "metadata": {},
   "source": [
    "We can use the LocusID to join our alleles to the locus table to create an expanded view of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "756b7de4-fa67-4857-92b4-ab1498e572bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>allele_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16682</td>\n",
       "      <td>16774</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16682</td>\n",
       "      <td>16774</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16682</td>\n",
       "      <td>16774</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16682</td>\n",
       "      <td>16774</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19275</td>\n",
       "      <td>19473</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213151</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248937524</td>\n",
       "      <td>248938705</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213152</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248937524</td>\n",
       "      <td>248938705</td>\n",
       "      <td>8588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213153</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248937524</td>\n",
       "      <td>248938705</td>\n",
       "      <td>8652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213154</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248937524</td>\n",
       "      <td>248938705</td>\n",
       "      <td>5149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213155</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248937524</td>\n",
       "      <td>248938705</td>\n",
       "      <td>5462</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213156 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end  allele_length\n",
       "0       chr1      16682      16774            165\n",
       "1       chr1      16682      16774            139\n",
       "2       chr1      16682      16774            163\n",
       "3       chr1      16682      16774            167\n",
       "4       chr1      19275      19473             70\n",
       "...      ...        ...        ...            ...\n",
       "213151  chr1  248937524  248938705             76\n",
       "213152  chr1  248937524  248938705           8588\n",
       "213153  chr1  248937524  248938705           8652\n",
       "213154  chr1  248937524  248938705           5149\n",
       "213155  chr1  248937524  248938705           5462\n",
       "\n",
       "[213156 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "view = allele.join(loci, on=\"LocusID\", rsuffix='_')\n",
    "view[[\"chrom\", \"start\", \"end\", \"allele_length\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b2b2f68-9ace-4221-9e5f-064d808673f0",
   "metadata": {},
   "source": [
    "What if we wanted to get the coordinates of loci with at least four alleles? We can use our `alleles_per_locus` created above to subset to loci with at least four alleles and then we'd again use the LocusID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "84019875-6a3d-4c2b-9c0d-145615f175f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of the 77213 Loci, 18218 have >= 4 alleles\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54682</td>\n",
       "      <td>54847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>27</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66127</td>\n",
       "      <td>66662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>23</td>\n",
       "      <td>chr1</td>\n",
       "      <td>98969</td>\n",
       "      <td>99141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>25</td>\n",
       "      <td>chr1</td>\n",
       "      <td>104130</td>\n",
       "      <td>104226</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    LocusID chrom   start     end\n",
       "4         0  chr1   30824   30989\n",
       "11       12  chr1   54682   54847\n",
       "13       27  chr1   66127   66662\n",
       "24       23  chr1   98969   99141\n",
       "26       25  chr1  104130  104226"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_subset = alleles_per_locus[alleles_per_locus >= 4].index # This Series' index is LocusID\n",
    "view = loci[loci[\"LocusID\"].isin(m_subset)]\n",
    "print(f\"Of the {len(loci)} Loci, {len(view)} have >= 4 alleles\")\n",
    "view.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd365e9-18ae-4068-a6af-91d0402a3783",
   "metadata": {},
   "source": [
    "Doing this `.isin()` operation isn't our only option. We can leverage pandas DataFrame indexing by setting the locus table's index to LocusID and then repeat the query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c244d16b-1f30-42f3-ab7a-2008fdcee03a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of the 77213 Loci, 18218 have >= 4 alleles\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>72379</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236405695</td>\n",
       "      <td>236405952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41145</th>\n",
       "      <td>chr1</td>\n",
       "      <td>119251051</td>\n",
       "      <td>119251288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69246</th>\n",
       "      <td>chr1</td>\n",
       "      <td>224011660</td>\n",
       "      <td>224016583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23149</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57421862</td>\n",
       "      <td>57421951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2075</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5977818</td>\n",
       "      <td>5977961</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end\n",
       "72379  chr1  236405695  236405952\n",
       "41145  chr1  119251051  119251288\n",
       "69246  chr1  224011660  224016583\n",
       "23149  chr1   57421862   57421951\n",
       "2075   chr1    5977818    5977961"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci = loci.set_index(\"LocusID\")\n",
    "view = loci.loc[m_subset]\n",
    "print(f\"Of the {len(loci)} Loci, {len(view)} have >= 4 alleles\")\n",
    "view.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27e3d384-cf1a-4934-adc7-3354f25835bf",
   "metadata": {},
   "source": [
    "## The Sample tables\n",
    "\n",
    "Every sample contained in a TRGTdb has its own table. Therefore, the `data` dictionary's key \"sample\" holds a value of a sub-dictionary with keys of the sample's name and a value with the DataFrame. We can see that in this example data, we have three samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f96cc5c9-b93c-4ce2-8323-c23f65271077",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['son', 'father', 'mother'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusID</th>\n",
       "      <th>allele_number</th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>135</td>\n",
       "      <td>167</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>69</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>106</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>94</td>\n",
       "      <td>95</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   LocusID  allele_number  spanning_reads  length_range_lower  \\\n",
       "0        0              1              15                 135   \n",
       "1        1              1              18                  69   \n",
       "2        2              1              18                 106   \n",
       "3        3              0              29                  90   \n",
       "4        4              0              65                  94   \n",
       "\n",
       "   length_range_upper  average_methylation  \n",
       "0                 167                  NaN  \n",
       "1                  70                  NaN  \n",
       "2                 109                  NaN  \n",
       "3                  95                  NaN  \n",
       "4                  95                  NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(data['sample'].keys())\n",
    "# And lets go ahead and look at the first sample's table\n",
    "data['sample']['son'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3440a878-4173-47e9-9fa3-b2eae55922f8",
   "metadata": {},
   "source": [
    "The columns of a sample table are:\n",
    "    \n",
    "- LocusID : The key that ties to the locus and allele table\n",
    "- allele_number : The key that ties to the allele table\n",
    "- spanning_reads : How many reads span the sample's observed allele\n",
    "- length_range_lower : Lower limit of the sample's observed allele length\n",
    "- length_range_upper : Upper limit of the sample's observed allele length\n",
    "- average_methylation : Sample's observed allele's mean methylation level\n",
    "\n",
    "We can check how many alleles are observed per-sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6524b431-521f-4799-a1af-8e9d5c776058",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "son has 154228 alleles\n",
      "father has 154134 alleles\n",
      "mother has 154186 alleles\n"
     ]
    }
   ],
   "source": [
    "for sample_name, table in data['sample'].items():\n",
    "    print(f\"{sample_name} has {len(table)} alleles\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f36a2c95-851f-40b9-a872-10fb24e17336",
   "metadata": {},
   "source": [
    "Or we can check how many alleles with at least 10 spanning reads are observed per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9abe4453-3bbf-4080-b17b-d0489e87bcda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "son has 140473 alleles with >= 10 spanning reads\n",
      "father has 125313 alleles with >= 10 spanning reads\n",
      "mother has 131477 alleles with >= 10 spanning reads\n"
     ]
    }
   ],
   "source": [
    "MINSPANCNT = 10\n",
    "for sample_name, table in data['sample'].items():\n",
    "    min_ten = table[table[\"spanning_reads\"] >= MINSPANCNT]\n",
    "    print(f\"{sample_name} has {len(min_ten)} alleles with >= {MINSPANCNT} spanning reads\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7315d3-dff7-4b43-b14c-079f5582d0ed",
   "metadata": {},
   "source": [
    "Another complexity to the sample tables (besides that there is one per-sample) is that we use the sample tables' LocusID and allele_number as keys. Therefore, when joining a sample table to the locus or allele tables, we need to consider which columns to use as keys.\n",
    "\n",
    "The simplier join is between locus and a sample since it only needs the LocusID column and above we already set the locus table's index to LocusID. For this example we'll look up the coordinates of loci over which our 'son' sample has a non-reference (i.e. alternate) allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "31384db7-9805-403c-ad12-f1da39d04600",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "son has 58555 loci with non-REF alleles\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LocusID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31777</td>\n",
       "      <td>31847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29714</td>\n",
       "      <td>29822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44805</td>\n",
       "      <td>44906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35458</td>\n",
       "      <td>35537</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  start    end\n",
       "LocusID                    \n",
       "0        chr1  30824  30989\n",
       "1        chr1  31777  31847\n",
       "2        chr1  29714  29822\n",
       "5        chr1  44805  44906\n",
       "6        chr1  35458  35537"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "son = data['sample']['son']\n",
    "\n",
    "son_nonref_alleles = son[son[\"allele_number\"] != 0]\n",
    "\n",
    "# We get the unique LocusID because the sample may have two different alternate alleles\n",
    "# such that a single LocusID would be in two rows e.g. (LocusID 1234 could have allele_number 1 and 2)\n",
    "son_nonref_loci = son_nonref_alleles[\"LocusID\"].unique() \n",
    "\n",
    "print(f\"son has {len(son_nonref_loci)} loci with non-REF alleles\")\n",
    "view = loci.loc[son_nonref_loci]\n",
    "view.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0cee94e-e1b8-4f9e-87d8-442aa5309509",
   "metadata": {},
   "source": [
    "The slightly more complex intersection is between the allele and sample tables. But after setting a MultiIndex on the allele table, the intersection is easy enough."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4146ef6b-9103-4725-9259-65981238de3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LocusID</th>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9654</th>\n",
       "      <th>1</th>\n",
       "      <td>76</td>\n",
       "      <td>TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19007</th>\n",
       "      <th>1</th>\n",
       "      <td>62</td>\n",
       "      <td>GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTAAAGTACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26033</th>\n",
       "      <th>1</th>\n",
       "      <td>139</td>\n",
       "      <td>AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56040</th>\n",
       "      <th>2</th>\n",
       "      <td>98</td>\n",
       "      <td>GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71960</th>\n",
       "      <th>2</th>\n",
       "      <td>76</td>\n",
       "      <td>GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTTCTTTTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       allele_length  \\\n",
       "LocusID allele_number                  \n",
       "9654    1                         76   \n",
       "19007   1                         62   \n",
       "26033   1                        139   \n",
       "56040   2                         98   \n",
       "71960   2                         76   \n",
       "\n",
       "                                                                sequence  \n",
       "LocusID allele_number                                                     \n",
       "9654    1              TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...  \n",
       "19007   1              GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTAAAGTACTT...  \n",
       "26033   1              AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...  \n",
       "56040   2              GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...  \n",
       "71960   2              GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTTCTTTTT...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = allele.set_index([\"LocusID\", \"allele_number\"])\n",
    "\n",
    "# Turning this into a list so we can use it to access the new alelle MultiIndex\n",
    "to_lookup = son_nonref_alleles[[\"LocusID\", \"allele_number\"]].values.tolist()\n",
    "son_sample = allele.loc[to_lookup].sample(5, random_state=42) # And we'll look at 5 random alleles.\n",
    "son_sample"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55bb12c2-2076-4bec-9f86-9ff89a2984d4",
   "metadata": {},
   "source": [
    "We could perform this intersection easier by simply merging the tables, though this comes at a cost to memory usage since we're creating a new, larger table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "90f34145-693b-4ce1-aab0-1e38e2857829",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusID</th>\n",
       "      <th>allele_number</th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTCTGTCTCCCTC...</td>\n",
       "      <td>15</td>\n",
       "      <td>135</td>\n",
       "      <td>167</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTCTGTCTCCCTC...</td>\n",
       "      <td>15</td>\n",
       "      <td>139</td>\n",
       "      <td>139</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>GTGTATATGCTCCACGATGCCTGTGAATATACAAACACACCACATC...</td>\n",
       "      <td>18</td>\n",
       "      <td>69</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "      <td>AGTGTATATGCTCCACGATGCCTGTGAATATACAAACACACCACAT...</td>\n",
       "      <td>6</td>\n",
       "      <td>71</td>\n",
       "      <td>73</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>GCCCTCGCCCTGGGTGCGCCCCGGGGCAGGACCCCCAGCCCAGGTC...</td>\n",
       "      <td>18</td>\n",
       "      <td>106</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   LocusID  allele_number  allele_length  \\\n",
       "0        0              1            139   \n",
       "1        0              1            139   \n",
       "2        1              1             70   \n",
       "3        1              2             71   \n",
       "4        2              1            108   \n",
       "\n",
       "                                            sequence  spanning_reads  \\\n",
       "0  TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTCTGTCTCCCTC...              15   \n",
       "1  TTATAGATCTGGTAAGAAAAGAGAGCATTGTCTCTCTGTCTCCCTC...              15   \n",
       "2  GTGTATATGCTCCACGATGCCTGTGAATATACAAACACACCACATC...              18   \n",
       "3  AGTGTATATGCTCCACGATGCCTGTGAATATACAAACACACCACAT...               6   \n",
       "4  GCCCTCGCCCTGGGTGCGCCCCGGGGCAGGACCCCCAGCCCAGGTC...              18   \n",
       "\n",
       "   length_range_lower  length_range_upper  average_methylation  \n",
       "0                 135                 167                  NaN  \n",
       "1                 139                 139                  NaN  \n",
       "2                  69                  70                  NaN  \n",
       "3                  71                  73                  NaN  \n",
       "4                 106                 109                  NaN  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "merged = pd.merge(allele.reset_index(), # unset the index we just made\n",
    "                  son_nonref_alleles, # then merge with this\n",
    "                  on=['LocusID','allele_number']) # on the two key columns\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4edc5e23-f818-473f-b7e4-17e573d37853",
   "metadata": {},
   "source": [
    "Inside our `son_sample` we have five non-REF alleles observed in the son. Let's check to see if those alleles are found in the mother and father samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ddeb93e5-a7e6-40ef-9f20-b5bf7af561bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LocusID</th>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19007</th>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56040</th>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>98</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       spanning_reads  length_range_lower  length_range_upper  \\\n",
       "LocusID allele_number                                                           \n",
       "19007   1                          14                  61                  62   \n",
       "56040   2                          15                  98                 100   \n",
       "\n",
       "                       average_methylation  \n",
       "LocusID allele_number                       \n",
       "19007   1                              NaN  \n",
       "56040   2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LocusID</th>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9654</th>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>75</td>\n",
       "      <td>76</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19007</th>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26033</th>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>137</td>\n",
       "      <td>141</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56040</th>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>96</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       spanning_reads  length_range_lower  length_range_upper  \\\n",
       "LocusID allele_number                                                           \n",
       "9654    1                          18                  75                  76   \n",
       "19007   1                          24                  61                  62   \n",
       "26033   1                          27                 137                 141   \n",
       "56040   2                          15                  96                 100   \n",
       "\n",
       "                       average_methylation  \n",
       "LocusID allele_number                       \n",
       "9654    1                              NaN  \n",
       "19007   1                              NaN  \n",
       "26033   1                              NaN  \n",
       "56040   2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mom = data['sample']['mother'].set_index([\"LocusID\", \"allele_number\"])\n",
    "display(mom[mom.index.isin(son_sample.index)])\n",
    "\n",
    "pop = data['sample']['father'].set_index([\"LocusID\", \"allele_number\"])\n",
    "display(pop[pop.index.isin(son_sample.index)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "083b7932-2cb6-48da-974c-d50df676886b",
   "metadata": {},
   "source": [
    "For these 5 non-reference alleles in the son, we can see that two of them from loci 19007 and 56040 are also present in both the parents. Two of them (LocusIDs 9654 and 26033) are found in only the father. And one of them (71960) isn't shared with either parent.\n",
    "\n",
    "Let's compare all of the alleles across these loci across all three samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d2689b38-3090-4c48-933e-29ad4b38ddb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "son = son.set_index([\"LocusID\", \"allele_number\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a3a0578f-d43c-46bc-a816-a795ec0c3414",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==== Examining Locus ====\n",
      "chrom        chr1\n",
      "start    23119226\n",
      "end      23119301\n",
      "Name: 9654, dtype: object\n",
      "All Alleles at the Locus:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>75</td>\n",
       "      <td>TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>76</td>\n",
       "      <td>TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>77</td>\n",
       "      <td>TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               allele_length  \\\n",
       "allele_number                  \n",
       "0                         75   \n",
       "1                         76   \n",
       "2                         77   \n",
       "\n",
       "                                                        sequence  \n",
       "allele_number                                                     \n",
       "0              TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...  \n",
       "1              TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...  \n",
       "2              TTTAACGACCTGATGTGACTAGCTCTTTTTTTTTTTTTTTTTTTTT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Son\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>75</td>\n",
       "      <td>76</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27</td>\n",
       "      <td>77</td>\n",
       "      <td>80</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "1                          14                  75                  76   \n",
       "2                          27                  77                  80   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "1                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mother\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>73</td>\n",
       "      <td>76</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>77</td>\n",
       "      <td>78</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "0                          18                  73                  76   \n",
       "2                          15                  77                  78   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "0                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Father\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>75</td>\n",
       "      <td>76</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>77</td>\n",
       "      <td>79</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "1                          18                  75                  76   \n",
       "2                           6                  77                  79   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "1                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==== Examining Locus ====\n",
      "chrom        chr1\n",
      "start    45032768\n",
      "end      45032831\n",
      "Name: 19007, dtype: object\n",
      "All Alleles at the Locus:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>63</td>\n",
       "      <td>GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTTAAAGTACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>62</td>\n",
       "      <td>GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTAAAGTACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>63</td>\n",
       "      <td>GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTAAAGTACTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               allele_length  \\\n",
       "allele_number                  \n",
       "0                         63   \n",
       "1                         62   \n",
       "2                         63   \n",
       "\n",
       "                                                        sequence  \n",
       "allele_number                                                     \n",
       "0              GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTTAAAGTACT...  \n",
       "1              GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTAAAGTACTT...  \n",
       "2              GAGCCACCGTGCCTGGCCTGATAATTTTTTTTTTTTTAAAGTACTT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Son\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>59</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "1                          27                  59                  62   \n",
       "2                           7                  63                  63   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "1                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mother\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>63</td>\n",
       "      <td>65</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "0                          19                  63                  65   \n",
       "1                          14                  61                  62   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "0                              NaN  \n",
       "1                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Father\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>63</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "0                          13                  63                  70   \n",
       "1                          24                  61                  62   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "0                              NaN  \n",
       "1                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==== Examining Locus ====\n",
      "chrom        chr1\n",
      "start    66545598\n",
      "end      66545725\n",
      "Name: 26033, dtype: object\n",
      "All Alleles at the Locus:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>139</td>\n",
       "      <td>AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>129</td>\n",
       "      <td>AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>129</td>\n",
       "      <td>AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               allele_length  \\\n",
       "allele_number                  \n",
       "0                        127   \n",
       "1                        139   \n",
       "2                        129   \n",
       "3                        129   \n",
       "\n",
       "                                                        sequence  \n",
       "allele_number                                                     \n",
       "0              AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...  \n",
       "1              AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...  \n",
       "2              AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...  \n",
       "3              AGTTCTCCAGAGAGACAGAACTGAACAAATTGTGTGTGTGTGTGTG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Son\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23</td>\n",
       "      <td>125</td>\n",
       "      <td>129</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>137</td>\n",
       "      <td>140</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "0                          23                 125                 129   \n",
       "1                          22                 137                 140   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "0                              NaN  \n",
       "1                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mother\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>125</td>\n",
       "      <td>128</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28</td>\n",
       "      <td>129</td>\n",
       "      <td>132</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "0                          30                 125                 128   \n",
       "2                          28                 129                 132   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "0                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Father\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>127</td>\n",
       "      <td>130</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>137</td>\n",
       "      <td>141</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "3                          18                 127                 130   \n",
       "1                          27                 137                 141   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "3                              NaN  \n",
       "1                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==== Examining Locus ====\n",
      "chrom         chr1\n",
      "start    187415702\n",
      "end      187415810\n",
      "Name: 56040, dtype: object\n",
      "All Alleles at the Locus:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>108</td>\n",
       "      <td>GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>96</td>\n",
       "      <td>GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98</td>\n",
       "      <td>GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104</td>\n",
       "      <td>GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               allele_length  \\\n",
       "allele_number                  \n",
       "0                        108   \n",
       "1                         96   \n",
       "2                         98   \n",
       "3                        104   \n",
       "\n",
       "                                                        sequence  \n",
       "allele_number                                                     \n",
       "0              GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...  \n",
       "1              GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...  \n",
       "2              GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...  \n",
       "3              GATTAGAACAATTGTGGCATAGTGGGGAAAACACACACACACACAC...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Son\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>95</td>\n",
       "      <td>97</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>98</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "1                          22                  95                  97   \n",
       "2                          18                  98                 100   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "1                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mother\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>93</td>\n",
       "      <td>97</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>98</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "1                          22                  93                  97   \n",
       "2                          15                  98                 100   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "1                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Father\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>96</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>102</td>\n",
       "      <td>106</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "2                          15                  96                 100   \n",
       "3                          22                 102                 106   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "2                              NaN  \n",
       "3                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==== Examining Locus ====\n",
      "chrom         chr1\n",
      "start    235572468\n",
      "end      235572544\n",
      "Name: 71960, dtype: object\n",
      "All Alleles at the Locus:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>76</td>\n",
       "      <td>GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTCTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>75</td>\n",
       "      <td>GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTCTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>76</td>\n",
       "      <td>GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTTCTTTTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               allele_length  \\\n",
       "allele_number                  \n",
       "0                         76   \n",
       "1                         75   \n",
       "2                         76   \n",
       "\n",
       "                                                        sequence  \n",
       "allele_number                                                     \n",
       "0              GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTCTTTTTT...  \n",
       "1              GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTCTTTTTT...  \n",
       "2              GGATTACAGGCATGAGCCACCGTGCTCTTTTTTTTTTTTTCTTTTT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Son\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>74</td>\n",
       "      <td>75</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>76</td>\n",
       "      <td>77</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "1                          30                  74                  75   \n",
       "2                          10                  76                  77   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "1                              NaN  \n",
       "2                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mother\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>76</td>\n",
       "      <td>77</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>74</td>\n",
       "      <td>75</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "0                          16                  76                  77   \n",
       "1                          18                  74                  75   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "0                              NaN  \n",
       "1                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Father\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spanning_reads</th>\n",
       "      <th>length_range_lower</th>\n",
       "      <th>length_range_upper</th>\n",
       "      <th>average_methylation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>76</td>\n",
       "      <td>77</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>73</td>\n",
       "      <td>75</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spanning_reads  length_range_lower  length_range_upper  \\\n",
       "allele_number                                                           \n",
       "0                           6                  76                  77   \n",
       "1                          21                  73                  75   \n",
       "\n",
       "               average_methylation  \n",
       "allele_number                       \n",
       "0                              NaN  \n",
       "1                              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for locus, allele_number in son_sample.index:\n",
    "    print(f\"==== Examining Locus ====\\n{loci.loc[locus]}\")\n",
    "    print(\"All Alleles at the Locus:\")\n",
    "    display(allele.loc[locus])\n",
    "    print(\"Son\")\n",
    "    display(son.loc[locus])\n",
    "    print(\"Mother\")\n",
    "    display(mom.loc[locus])\n",
    "    print(\"Father\")\n",
    "    display(pop.loc[locus])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2522370-ab3d-40a5-9c4d-b94295f439af",
   "metadata": {},
   "source": [
    "We'll leave it as an exercise for the reader to determine the mendelian consistency of the variants in these 5 loci. Also, you may find combining the tables presented above into a single report to be excellent practice for creating your own TRGTdb queries."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33d6d722-6b52-410c-95dd-25a49b3a143e",
   "metadata": {},
   "source": [
    "# TRGTdb queries\n",
    "\n",
    "The tutorial up to this point has really been more of a primer on how to use Pandas. However, it's possible to perform some queries on a TRGTdb without too much Pandas knowledge. Below we'll show you the results and descriptions of some of the basic TRGTdb queries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "277dce93-dc89-4f48-bf58-5a121045d7f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Allele counts and frequency\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>allele_number</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LocusID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31777</td>\n",
       "      <td>31847</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31777</td>\n",
       "      <td>31847</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77211</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248752443</td>\n",
       "      <td>248754675</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195904 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end  allele_number  AC        AF\n",
       "LocusID                                                         \n",
       "0        chr1      30824      30989              1   4  0.666667\n",
       "0        chr1      30824      30989              2   1  0.166667\n",
       "0        chr1      30824      30989              3   1  0.166667\n",
       "1        chr1      31777      31847              1   3  0.500000\n",
       "1        chr1      31777      31847              0   2  0.333333\n",
       "...       ...        ...        ...            ...  ..       ...\n",
       "77211    chr1  248752443  248754675              4   1  0.166667\n",
       "77212    chr1  248404165  248404241              1   2  0.333333\n",
       "77212    chr1  248404165  248404241              2   2  0.333333\n",
       "77212    chr1  248404165  248404241              3   1  0.166667\n",
       "77212    chr1  248404165  248404241              4   1  0.166667\n",
       "\n",
       "[195904 rows x 6 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is equivalent to the command line `trgt db query allele_cnts data.tdb`\n",
    "print(trgt.allele_count.__doc__)\n",
    "trgt.allele_count(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bb39aca0-f162-44ae-a2e5-5dd29dd9cbfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Allele counts and frequency by allele length\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>is_ref</th>\n",
       "      <th>allele_length</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LocusID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "      <td>False</td>\n",
       "      <td>139</td>\n",
       "      <td>4</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "      <td>False</td>\n",
       "      <td>163</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30824</td>\n",
       "      <td>30989</td>\n",
       "      <td>False</td>\n",
       "      <td>167</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31777</td>\n",
       "      <td>31847</td>\n",
       "      <td>True</td>\n",
       "      <td>70</td>\n",
       "      <td>5</td>\n",
       "      <td>0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31777</td>\n",
       "      <td>31847</td>\n",
       "      <td>False</td>\n",
       "      <td>71</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77211</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248752443</td>\n",
       "      <td>248754675</td>\n",
       "      <td>False</td>\n",
       "      <td>2227</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>False</td>\n",
       "      <td>8588</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>False</td>\n",
       "      <td>8652</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>False</td>\n",
       "      <td>5149</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77212</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248404165</td>\n",
       "      <td>248404241</td>\n",
       "      <td>False</td>\n",
       "      <td>5462</td>\n",
       "      <td>1</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>185155 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end is_ref  allele_length  AC        AF\n",
       "LocusID                                                                \n",
       "0        chr1      30824      30989  False            139   4  0.666667\n",
       "0        chr1      30824      30989  False            163   1  0.166667\n",
       "0        chr1      30824      30989  False            167   1  0.166667\n",
       "1        chr1      31777      31847   True             70   5  0.833333\n",
       "1        chr1      31777      31847  False             71   1  0.166667\n",
       "...       ...        ...        ...    ...            ...  ..       ...\n",
       "77211    chr1  248752443  248754675  False           2227   1  0.166667\n",
       "77212    chr1  248404165  248404241  False           8588   2  0.333333\n",
       "77212    chr1  248404165  248404241  False           8652   2  0.333333\n",
       "77212    chr1  248404165  248404241  False           5149   1  0.166667\n",
       "77212    chr1  248404165  248404241  False           5462   1  0.166667\n",
       "\n",
       "[185155 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is equivalent to the command line `trgt db query allele_cnts_bylen data.tdb`\n",
    "print(trgt.allele_count_length.__doc__)\n",
    "trgt.allele_count_length(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8c2b052c-0683-46e2-a3a1-9712b2c90c53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Monozygotic reference sites per-sample and overall\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>loci</th>\n",
       "      <th>mon_ref</th>\n",
       "      <th>pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>son</td>\n",
       "      <td>77114</td>\n",
       "      <td>18559</td>\n",
       "      <td>0.240670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>father</td>\n",
       "      <td>77067</td>\n",
       "      <td>18344</td>\n",
       "      <td>0.238027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mother</td>\n",
       "      <td>77093</td>\n",
       "      <td>17574</td>\n",
       "      <td>0.227958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>all</td>\n",
       "      <td>77213</td>\n",
       "      <td>11040</td>\n",
       "      <td>0.142981</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample   loci  mon_ref       pct\n",
       "0     son  77114    18559  0.240670\n",
       "1  father  77067    18344  0.238027\n",
       "2  mother  77093    17574  0.227958\n",
       "3     all  77213    11040  0.142981"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is equivalent to the command line `trgt db query monref data.tdb`\n",
    "print(trgt.monref.__doc__)\n",
    "trgt.monref(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "06ec34d4-2e49-4b68-9de6-eb555ca44c4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Allele length, methylation, and CpG stats (PMID:3656447)\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusID</th>\n",
       "      <th>allele_number</th>\n",
       "      <th>allele_length</th>\n",
       "      <th>average_methylation</th>\n",
       "      <th>CpG_obs</th>\n",
       "      <th>CpG_exp</th>\n",
       "      <th>CpG_density</th>\n",
       "      <th>GC_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5.381295</td>\n",
       "      <td>0.014388</td>\n",
       "      <td>0.438849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5.381295</td>\n",
       "      <td>0.014388</td>\n",
       "      <td>0.438849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>3.257143</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.442857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>3.211268</td>\n",
       "      <td>0.028169</td>\n",
       "      <td>0.436620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>0.092593</td>\n",
       "      <td>0.740741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154181</th>\n",
       "      <td>77210</td>\n",
       "      <td>3</td>\n",
       "      <td>902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>32.258315</td>\n",
       "      <td>0.028825</td>\n",
       "      <td>0.379157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154182</th>\n",
       "      <td>77211</td>\n",
       "      <td>2</td>\n",
       "      <td>2906</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>25.569855</td>\n",
       "      <td>0.007571</td>\n",
       "      <td>0.500688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154183</th>\n",
       "      <td>77211</td>\n",
       "      <td>3</td>\n",
       "      <td>2227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>23.278850</td>\n",
       "      <td>0.007185</td>\n",
       "      <td>0.497081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154184</th>\n",
       "      <td>77212</td>\n",
       "      <td>2</td>\n",
       "      <td>8652</td>\n",
       "      <td>NaN</td>\n",
       "      <td>328</td>\n",
       "      <td>341.391586</td>\n",
       "      <td>0.075821</td>\n",
       "      <td>0.438165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154185</th>\n",
       "      <td>77212</td>\n",
       "      <td>3</td>\n",
       "      <td>5149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>196</td>\n",
       "      <td>200.241600</td>\n",
       "      <td>0.076131</td>\n",
       "      <td>0.436007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>462548 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        LocusID  allele_number  allele_length  average_methylation  CpG_obs  \\\n",
       "0             0              1            139                  NaN        1   \n",
       "1             0              1            139                  NaN        1   \n",
       "2             1              1             70                  NaN        1   \n",
       "3             1              2             71                  NaN        1   \n",
       "4             2              1            108                  NaN        5   \n",
       "...         ...            ...            ...                  ...      ...   \n",
       "154181    77210              3            902                  NaN       13   \n",
       "154182    77211              2           2906                  NaN       11   \n",
       "154183    77211              3           2227                  NaN        8   \n",
       "154184    77212              2           8652                  NaN      328   \n",
       "154185    77212              3           5149                  NaN      196   \n",
       "\n",
       "           CpG_exp  CpG_density    GC_pct  \n",
       "0         5.381295     0.014388  0.438849  \n",
       "1         5.381295     0.014388  0.438849  \n",
       "2         3.257143     0.028571  0.442857  \n",
       "3         3.211268     0.028169  0.436620  \n",
       "4        13.000000     0.092593  0.740741  \n",
       "...            ...          ...       ...  \n",
       "154181   32.258315     0.028825  0.379157  \n",
       "154182   25.569855     0.007571  0.500688  \n",
       "154183   23.278850     0.007185  0.497081  \n",
       "154184  341.391586     0.075821  0.438165  \n",
       "154185  200.241600     0.076131  0.436007  \n",
       "\n",
       "[462548 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is equivalent to the command line `trgt db query methyl data.tdb`\n",
    "print(trgt.methyl.__doc__)\n",
    "trgt.methyl(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb6ed675-27f1-42e7-8bfd-782560eab0b7",
   "metadata": {},
   "source": [
    "# Next Steps\n",
    "\n",
    "From here, you're ready to jump into some advanced querying *notebooklink*\n",
    "Alternatively, you may be interested in how efficiently load subsets of the the TRGTdb *notebooklink*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6103f4e-5b22-4e25-8f76-b16b81765574",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
