# Command-line interface of TRGT and TRVZ

## TRGTs command-line options

- `--genome <GENOME>` Path to the FASTA file containing reference genome. This
  must be the same reference genomes as the one used for read alignment.
- `--reads <READS>` BAM file with alignments of HiFi reads.
- `--repeats <REPEATS>`  BED file with reference coordinates and structure of
  tandem repeats.
- `--output-prefix <OUTPUT_PREFIX>` Prefix for output files. TRGT generates an
  unsorted VCF file (`<OUTPUT_PREFIX>.vcf.gz`) and unsorted BAM file with pieces
  of HiFi reads overlapping the repeats (`<OUTPUT_PREFIX>.spanning.bam`).
- `--threads <THREADS>` Number of threads. Set to 1 by default.
- `--karyotype <KARYOTYPE>` Sample karyotype (XX or XY). Set to XX by default.
- `--sample-name <SAMPLE_NAME>` Sample name. If it is not provided, the sample name is extracted from the input BAM or file stem.
- `--max-depth <MAX_DEPTH>` Maximum locus depth. Set to 250 by default.

## TRVZ command-line options

- `--genome <GENOME>` Path to the FASTA file containing reference genome.
- `--repeats <REPEATS>` BED file with repeat coordinates and structure.
- `--spanning-reads <SPANNING_READS>` BAM file generated by TRGT.
- `--vcf <VCF>` VCF file generated by TRGT.
- `--repeat-id <REPEAT_ID>` ID of the repeat to visualize.
- `--plot-type <plot type>` TRVZ can generate two types of plots: allele plots
  and waterfall plots. Allele plots show alignments of reads to each repeat
  allele. Waterfall plots display unaligned repeat sequences without aligning
  them to the (consensus) allele. Waterfall plots are especially useful for QC
  of repeat calls and for visualization of mosaic expansions. TRVZ generates
  allele plots by default.
- `--show <show>` TRVZ can either visualize motifs (`motifs`) or methylation
  (`meth`). Motifs are visualized by default.
- `--image <OUTPUT_PATH>` Output image path with ends with extension .pdf, .svg,
  or .png.
